04.ls/ls.rb

```diff
@@ -1,15 +1,97 @@
 #!/usr/bin/env ruby
 # frozen_string_literal: true
 
+require 'etc'
+
 require 'optparse'
 
 SPLIT_NUMBER = 3
 
+FILE_TYPE = {
+  '01' => 'p',
+  '02' => 'c',
+  '04' => 'd',
+  '06' => 'b',
+  '10' => '-',
+  '12' => 'l',
+  '14' => 's'
+}.freeze
+
+FILE_MODE = {
+  '0' => '---',
+  '1' => '--x',
+  '3' => '-w-',
+  '4' => 'r--',
+  '5' => 'r-x',
+  '6' => 'rw-',
+  '7' => 'rwx'
+}.freeze
+
 def main
-  options = ARGV.getopts('r')
+  options = ARGV.getopts('l')
   files = Dir.glob('*')
-  sorted_files = options['r'] ? files.reverse : files
-  display_files(sorted_files)
+  options['l'] ? display_long_format(files) : display_files(files)
+end
+
+def display_long_format(files)
+  long_formats = gain_long_format(files)
+  total_block_number = calc_total_block_number(long_formats)
+  max_width = calc_max_width(long_formats)
+
+  puts "total #{total_block_number}"
+  long_formats.each do |format|
+    print "#{format[:file_mode]}  "
+    print "#{format[:hard_links_count].to_s.rjust(max_width[:link])} "
+    print "#{format[:owner].ljust(max_width[:owner])}  "
+    print "#{format[:group].ljust(max_width[:group])}  "
+    print "#{format[:file_size].to_s.rjust(max_width[:file_size])} "
+    print "#{format[:day_time]} "
+    print "#{format[:file_name]}\n"
+  end
+end
+
+def gain_long_format(files)
+  files.map do |file|
+    file_info = File.stat(file)
+    {
+      block: file_info.blocks,
+      file_mode: fetch_file_mode(file_info),
+      hard_links_count: file_info.nlink,
+      owner: Etc.getpwuid(file_info.uid).name,
+      group: Etc.getgrgid(file_info.gid).name,
+      file_size: file_info.size,
+      day_time: file_info.mtime.strftime('%_m %_d %H:%M'),
+      file_name: File.basename(file)
+    }
+  end
+end
+
+def calc_total_block_number(long_formats)
+  block_long_formats = long_formats.map { |long_format| long_format.fetch(:block) }
+  block_long_formats.sum
```

Comment:
```suggestion
  long_formats.sum { |long_format| long_format.fetch(:block) }
```

これだけでいけます

ちなみに`[]`ではなく`fetch`を使う理由って何かありますか？
（`[]`の方が短くていいような）

Reply:
fetchだと例外処理を出力してくれるので[]より適切かな?と思い使用しました。
しかし、よくよく考えてみるとキー自体はユーザーが指定することはなく、例外処理を発生させたい場面がそもそもないので[]に修正したいと思います🙇